Editorial review (#80)
iov-one/docs: docs/weave/weave-api-spec/01-address-derivation.md

---
id: address-derivation 
title: Addresses Derivation
sidebar_label:  Address Derivation
---

IOV uses industry standard technology to derive cryptographic key pairs and
blockchain addresses from a secret mnemonic. The first step is to derive an
Ed25519 key pair from the secret. The address is then derived from the public
key in a second step.

## Create the base secret

The base secret is an English
[BIP39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki)-compatible
mnemonic.

Even though it is potentially feasible to extend this spec for the use of
mnemonics in different languages, none of our software currently supports doing
that.

Note: There are extensions to BIP39 that allow mnemonics with fewer than 12

> words. Those mnemonics should be considered insecure for our purposes and none

Reviewer: ```suggestion
words. Those mnemonics should be considered insecure for our purposes, and none
```

of our software supports them.

## Derive a key pair from a mnemonic

The IOV Name Service (i.e. our blockchain) uses Ed25519 for signing.
Hierarchical deterministic (HD) key derivation allows us to derive multiple
Ed25519 key pairs from a single mnemonic. We follow the
[SLIP-0010](https://github.com/satoshilabs/slips/blob/master/slip-0010.md)
standard.

## Derivation paths

[BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki)
describes a HD derivation scheme that allows using multiple coins with a single
mnemonic. Following this standard, we registered the coin index 234 for the IOV
token in
[SLIP-0044](https://github.com/satoshilabs/slips/blob/master/slip-0044.md).

Since IOV is an account-based model, our paths follows
[Trezor's recommendation](https://github.com/trezor/trezor-firmware/tree/master/core/docs/coins)
that

> If the coin is UTXO-based the path should have all five parts, precisely as
> defined in BIP-32. If it is account-based we follow Stellar's SEP-0005 - paths
> have only three parts 44'/c'/a'.

This results in HD paths of the form

```
m/44'/234'/a'
```

> where `a` is an account index starting at 0. We call this format "BIP44-like"

Reviewer: ```suggestion
where `a` is an account index starting at 0. We call this format "BIP44-like"
```
```suggestion
where `a` is an account index starting at 0. We call this format 'BIP44-like'
```

> because it follows the idea of BIP44 but is not compatible to the 5-component

Reviewer: ```suggestion
because it follows the idea of BIP44, but is not compatible with the five-component
```

BIP44 standard.

### Simple addresses (deprecated)

[During development](https://github.com/iov-one/iov-core/blob/v0.15.0/docs/KeyBase.md#simple-addresses),
we used the derivation path `m/4804438'/a'`. This was created and called

> "simple" because we initially planned to use a multi-dimensional 5-component

Reviewer: ```suggestion
'simple' because we initially planned to use a multi-dimensional five-component
```

> BIP44 path in the long run. However, we gave up this idea and now the IOV path

Reviewer: ```suggestion
BIP44 path in the long run. However, we gave up on this idea and now the IOV path
```
